realm/realm-library/src/objectServer/java/io/realm/SyncManager.java

```diff
@@ -230,8 +227,57 @@ static void notifyUserLoggedOut(SyncUser user) {
         }
     }
 
+    // This is called from SyncManager.cpp from the worker thread the Sync Client is running on
+    // Right now Core doesn't send these errors to the proper session, so instead we need to notify all sessions
+    // from here. This can be removed once better error propagation is implemented in Sync Core.
+
+    /**
+     * All errors from native Sync is reported to this method. From the path we can determine which
+     * session to contact. If {@code path == null} all sessions are effected.
+     */
+    @SuppressWarnings("unused")
+    private static synchronized void notifyErrorHandler(int errorCode, String errorMessage, String path) {
+        for (SyncSession syncSession : sessions.values()) {
+            if (path == null || path.equals(syncSession.getConfiguration().getPath())) {
+                syncSession.notifySessionError(errorCode, errorMessage);
```

Comment:
After discussions with @beeender, I believed we agreed that we need to catch all exceptions in callbacks from the OS and log them to logcat instead of throwing them, since it is close to impossible to make C++ terminate correctly for the Java session to actually terminate the app (instead of being swallowed in JNI causing other crashes). Also other libraries like RxJava have similar policies.